File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
from collections import deque

def BFS(start_id, visited, computers):
    queue = deque()
    queue.append(start_id)
    visited[start_id] = True
    
    while queue:
        cur_id = queue.popleft()
        
        for new_id, value in enumerate(computers[cur_id]):
            if new_id != cur_id and not visited[new_id] and value == 1:
                queue.append(new_id)
                visited[new_id] = True
    
    

def solution(n, computers):
    answer = 0
    visited = [False for _ in range(n)]
    for id in range(n):
        if not visited[id]:
            BFS(id, visited, computers)
            answer += 1
    return answer

File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    boolean[] visited;
    Queue<Integer> q = new LinkedList<Integer>();
    int answer = 0;
      
    public int solution(int n, int[][] computers) {
        visited = new boolean[n];
        
        for(int i=0;i<n;i++){
            if(!visited[i]) {
                bfs(i, computers, n);
                answer++;
            } 
        }
        return answer;
    }
    
    public void bfs(int start, int[][] computers, int n) {
        q.offer(start);
        visited[start] = true;
        
        while(!q.isEmpty()) {
            int node = q.poll();
            for (int next=0; next< n; next++) {
                if (!visited[next] && computers[node][next] == 1) {
                    q.offer(next);
                    visited[next] = true;
                }
            }
        }
    }
}